proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+               })
+    )
+
+refund: _ => _
+
+requires:
+    andBool 0 <=Int TO_ID     andBool TO_ID     <Int (2 ^Int 160)
+    andBool 0 <=Int VALUE     andBool VALUE     <Int (2 ^Int 256)
+    andBool 0 <=Int BAL_FROM  andBool BAL_FROM  <Int (2 ^Int 256)
+    andBool 0 <=Int BAL_TO    andBool BAL_TO    <Int (2 ^Int 256)
+
+comments:
+    // NOTE: negative VALUE is not possible since it is of `num256` type
+```
+
+There are two sub-cases depending on whether the caller is equal to the recipient.
+
+#### transfer success: sub-case 1: the caller `CALLER_ID` is different from the recipient `TO_ID`.
+
+```{.k .erc20-spec .hkg-spec}
+[transfer-success-1]
+```
+
+`<storage>` cell specifies that the amount of `VALUE` is transferred from `balances[CALLER_ID]` to `balances[TO_ID]`.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(balances)}, CALLER_ID) |-> (BAL_FROM => BAL_FROM -Int VALUE)
+```
+
+```{.k .erc20-spec}
+    #hashedLocation({lang}, {pos(balances)}, TO_ID)     |-> (BAL_TO   => BAL_TO   +Int VALUE)
+```
+
+##### HKG-specific:
+On the other hand, HKG token allows the additional overflow, “wrapping around” the balance if it is more than `2^256`, which is very unlikely but still a questionable behavior. See the difference where `+Word`, the addition modulo `2^256`, is used instead of `+Int`, the mathematical, arbitrary precision integer addition.
+
+```{.k .hkg-spec}
+    #hashedLocation({lang}, {pos(balances)}, TO_ID)     |-> (BAL_TO   => BAL_TO  +Word VALUE)
+```
+
+```{.k .erc20-spec .hkg-spec}
+    _:Map
+
++requires:
+    andBool CALLER_ID =/=Int TO_ID
+```
+
+The side-condition ensures that no arithmetic overflow happens.
+
+```{.k .erc20-spec}
+    andBool VALUE <=Int BAL_FROM
+    andBool BAL_TO +Int VALUE <Int (2 ^Int 256) // viper overflow check: (VALUE ==Int 0 xorBool BAL_TO +Word VALUE >Int BAL_TO)
+```
+
+##### HKG-specific:
+As mentioned earlier, HKG token does not check the overflow thus we have no side-condition for that. Instead, HKG token does not allow the zero value to be requested to transfer, which is questionable though, and we have the side-condition for that.
+
+```{.k .hkg-spec}
+    andBool VALUE <=Int BAL_FROM
+    andBool VALUE >Int 0
+```
+
+
+#### transfer success: sub-case 2: the caller `CALLER_ID` is equal to the recipient `TO_ID`.
+
+```{.k .erc20-spec .hkg-spec}
+[transfer-success-2]
+```
+
+`<storage>` cell specifies that `balances[CALLER_ID]` is not changed since the value is transferred to himself.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(balances)}, CALLER_ID) |-> BAL_FROM
+    _:Map
+
++requires:
+    andBool CALLER_ID ==Int TO_ID
+```
+
+The side-condition ensures that no arithmetic overflow (precisely underflow in this case) happens.
+
+```{.k .erc20-spec}
+    andBool VALUE <=Int BAL_FROM
+```
+
+##### HKG-specific:
+As in the sub-case 1, HKG token requires having the additional side-condition of VALUE’s being greater than 0.
```

Comment:
I find it very confusing to present both the correct ERC20 variant and the HKG token variant at the same time, in parallel.  I would present each of them separately, and when presented the HKG token I would say that there are some differences and make those differences very clearly marked.  Each token is likely to be be somewhat different in one way or another.  So what are we going to do, to include all of them in the same file, and then lose track of the slight differences and where is one and where is the other?

Reply:
I felt similarly once I wrote it. I did it in this way because I thought we can avoid any duplication of the spec, which helps to maintain it. But at this point, I think it'd be better to split the spec into two files and extend the inheritance figure, minimizing duplication in the second file.

Reply:
Can you separate them but still avoid duplication?  Introduce some notation in your DSL that you are extending another module, and then continue to use your overriding.